routers/repo/compare.go

```diff
@@ -139,9 +139,47 @@ func ParseCompareInfo(ctx *context.Context) (*models.User, *models.Repository, *
 	ctx.Data["BaseIsBranch"] = baseIsBranch
 	ctx.Data["BaseIsTag"] = baseIsTag
 
+	var forkedRepo *models.Repository
+	if baseRepo.IsFork {
+		err = baseRepo.GetBaseRepo()
+		if err != nil {
+			if !models.IsErrRepoNotExist(err) {
+				ctx.ServerError("Unable to find root repo", err)
+				return nil, nil, nil, nil, "", ""
+			}
+		} else {
+			forkedRepo = baseRepo.BaseRepo
+			ctx.Data["ForkedRepo"] = forkedRepo
+		}
+	}
+
+	var ownForkedRepo *models.Repository
+	if baseRepo.OwnerID != ctx.User.ID {
+		repo, has := models.HasForkedRepo(ctx.User.ID, baseRepo.ID)
+		if has {
+			ownForkedRepo = repo
+			ctx.Data["OwnForkedRepo"] = ownForkedRepo
+		}
+	}
+
+	headRepo := baseRepo
+	has := false
 	// Check if current user has fork of repository or in the same repository.
-	headRepo, has := models.HasForkedRepo(headUser.ID, baseRepo.ID)
-	if !has && !isSameRepo {
+	if !isSameRepo && forkedRepo != nil && forkedRepo.OwnerID == headUser.ID {
+		headRepo = forkedRepo
+		has = true
+	}
+	if !isSameRepo && ownForkedRepo != nil && ownForkedRepo.OwnerID == headUser.ID {
+		headRepo = ownForkedRepo
+		has = true
+	}
+	if !isSameRepo && !has {
+		headRepo, has = models.HasForkedRepo(headUser.ID, baseRepo.ID)
+	}
+	if !isSameRepo && !has && baseRepo.IsFork {
+		headRepo, has = models.HasForkedRepo(headUser.ID, baseRepo.ForkID)
+	}
+	if !isSameRepo && !has {
```

Comment:
I'm very confused about what are you trying to accomplish here. You're comparing user IDs but assigning repositories? 🤔 Could you please top each `if` with comments about what cases are they covering?